src/storage/src/manifest/layer.rs

```diff
@@ -0,0 +1,82 @@
+use std::fmt::Debug;
+
+use async_trait::async_trait;
+use common_datasource::compression::CompressionType;
+
+use crate::error::{CompressObjectSnafu, DecompressObjectSnafu, Result};
+
+#[async_trait]
+pub trait ObjectStoreLayer: Send + Sync + Debug {
+    async fn forward(&self, data: Vec<u8>) -> Result<Vec<u8>>;
+    async fn backward(&self, data: Vec<u8>) -> Result<Vec<u8>>;
+}
+
+/// Layer used to compress data
+#[derive(Debug, Clone)]
+pub struct CompressLayer {
+    compress_type: CompressionType,
+}
+
+impl CompressLayer {
+    pub fn new(compress_type: CompressionType) -> Self {
+        Self { compress_type }
+    }
+}
+
+#[async_trait]
+impl ObjectStoreLayer for CompressLayer {
+    async fn forward(&self, data: Vec<u8>) -> Result<Vec<u8>> {
+        let compress_data = self.compress_type.encode(&data).await.map_err(|e| {
+            CompressObjectSnafu {
+                compress_type: self.compress_type,
+                err_msg: e.to_string(),
+            }
+            .build()
+        })?;
```

Comment:
Once you change `err_msg` to `source`, you can simply use context:
```suggestion
        let compress_data =
            self.compress_type
                .encode(&data)
                .await
                .context(CompressObjectSnafu {
                    compress_type: self.compress_type,
                    err_msg: e.to_string(),
                })?;
```